Use comment status and type to set $maybe_notify default

adamsilverstein opened the pull request:
<!--
Hi there! Thanks for contributing to WordPress!

Pull Requests in this GitHub repository **must** be linked to a ticket in the WordPress Core Trac instance (https://core.trac.wordpress.org), and are only used for code review. **No pull requests will be merged on GitHub.**

See the WordPress Handbook page on using PRs for Code Review more information: https://make.wordpress.org/core/handbook/contribute/git/github-pull-requests-for-code-review/

If this is your first time contributing, you may also find reviewing these guides first to be helpful:
- FAQs for New Contributors: https://make.wordpress.org/core/handbook/tutorials/faq-for-new-contributors/
- Contributing with Code Guide: https://make.wordpress.org/core/handbook/contribute/
- WordPress Coding Standards: https://make.wordpress.org/core/handbook/best-practices/coding-standards/
- Inline Documentation Standards: https://make.wordpress.org/core/handbook/best-practices/inline-documentation-standards/
- Browser Support Policies: https://make.wordpress.org/core/handbook/best-practices/browser-support/
- Proper spelling and grammar related best practices: https://make.wordpress.org/core/handbook/best-practices/spelling/
-->

<!-- Insert a description of your changes here -->

Trac ticket: https://core.trac.wordpress.org/ticket/64217
<!-- insert a link to the WordPress Trac ticket here -->

---
**This Pull Request is for code review only. Please keep all other discussion in the Trac ticket. Do not merge this Pull Request. See [GitHub Pull Requests for Code Review](https://make.wordpress.org/core/handbook/contribute/git/github-pull-requests-for-code-review/) in the Core Handbook for more details.**


---

## Commit message

```
Comments: Check comment approval before the `notify_post_author` filter.

The `notify_post_author` filter is documented as overriding the site setting, but the comment approval and type checks ran after the filter, so a callback returning true could not force a notification for an unapproved comment.

Move those checks before the filter by using the comment status and type to build the default value: notes default to the `wp_notes_notify` option regardless of approval status, unapproved comments default to false, and approved comments default to the `comments_notify` option. The filter now receives false for unapproved comments instead of the raw option value, giving developers complete control over notifications. Invalid comment IDs now return false without firing the filter.

Props westonruter, jorbin, audrasjb.
Fixes #64217.
```

adamsilverstein committed:
Use comment status and type to set $maybe_notify default

adamsilverstein committed:
Update filter doc block

westonruter commented on src/wp-includes/comment.php:
```diff
@@ -2429,9 +2429,17 @@ function wp_new_comment_notify_postauthor( $comment_id ) {
+		! isset( $comment->comment_approved ) ||
+		( '1' !== $comment->comment_approved && ! $is_note ) ) {
+			$maybe_notify = false;
+	}
```
Maybe this refactor into a single `if/elseif/else` statement is easier to reason about?

```suggestion
	// By default, only notify for approved comments and notes.
	if (
		! isset( $comment->comment_approved ) ||
		( '1' !== $comment->comment_approved && ! $is_note ) 
	) {
		$maybe_notify = false;
	} else if ( $is_note ) {
		$maybe_notify = get_option( 'wp_notes_notify' );
	} else {
		$maybe_notify = get_option( 'comments_notify' )
	}
```

westonruter replied:
Aside: Will `$comment->comment_approved` ever be not-set or `null`? It's default value is `'1'` and it is typed as a `string`. Doesn't hurt, but just curious. If so, then `! isset( $comment->comment_approved )` could be removed.

adamsilverstein replied:
This will change the logic slightly. If get_option( 'comments_notify' ) is true and '1' !== $comment->comment_approved and the type is a comment, the default should be false. This would make it true. I will try to make the logic clearer, maybe remove the trinary or add inline docs.

westonruter replied:
I may have just provided a poor suggestion, so don't feel like it needs to be actioned on 😊

aaronjorbin commented on src/wp-includes/comment.php:
```diff
@@ -2429,9 +2429,17 @@ function wp_new_comment_notify_postauthor( $comment_id ) {
+	 * overriding the site settings and defaults. By default, notifications are sent for
+	 * all notes and for approved comments.
 	 *
 	 * @since 4.4.0
```
```suggestion
	 * @since 4.4.0
	 * @since 6.9.0 Comment approval status is checked before this filter. 
```

Minor, but I think it's worth documenting the change for posterity.

adamsilverstein replied:
Thanks for the suggestion! I'll change this to 7.0 as I plan to hold off on this logic change so late in the release cycle.

adamsilverstein committed:
Update comment.php

Co-authored-by: Aaron Jorbin <aaronjorbin@users.noreply.github.com>

adamsilverstein committed:
Apply suggestion from @westonruter

Co-authored-by: Weston Ruter <westonruter@gmail.com>

westonruter commented on src/wp-includes/comment.php:
```diff
@@ -2474,13 +2474,25 @@ function wp_new_comment_notify_postauthor( $comment_id ) {
+	if (
+		! isset( $comment->comment_approved ) ||
+		( '1' !== $comment->comment_approved && ! $is_note )
+	) {
```
Is the `isset()` check needed? The `\WP_Comment::$comment_approved` variable is defined on `WP_Comment` as a string with a default value of `'1'`.

```suggestion
	if ( '1' !== $comment->comment_approved && ! $is_note ) {
```

Nevertheless, I see this is existing code from below (and maybe I commented this before).

westonruter commented on src/wp-includes/comment.php:
```diff
@@ -2474,13 +2474,25 @@ function wp_new_comment_notify_postauthor( $comment_id ) {
 	$comment = get_comment( $comment_id );
```
```suggestion
	$comment = get_comment( $comment_id );
	if ( ! ( $comment instanceof WP_Comment ) ) {
		return false;
	}
```

adamsilverstein committed:
Apply suggestions from Weston

Co-authored-by: Weston Ruter <westonruter@gmail.com>

adamsilverstein committed:
Merge branch 'trunk' into fix/notification-filter-default

westonruter commented on src/wp-includes/comment.php:
```diff
@@ -2472,15 +2472,27 @@ function wp_new_comment_notify_moderator( $comment_id ) {
+	if ( ! ( $comment instanceof WP_Comment ) ) {
+		return false;
+	}
 	$is_note = ( $comment && 'note' === $comment->comment_type );
```
Thanks to the above, this can now be simplified:

```suggestion
	$is_note = ( 'note' === $comment->comment_type );
```

westonruter commented on src/wp-includes/comment.php:
```diff
@@ -2472,15 +2472,27 @@ function wp_new_comment_notify_moderator( $comment_id ) {
  */
 function wp_new_comment_notify_postauthor( $comment_id ) {
 	$comment = get_comment( $comment_id );
+	if ( ! ( $comment instanceof WP_Comment ) ) {
```
Note: hereafter in this function, it should really use `$comment->comment_ID` (or whatever the field is) instead of `$comment_id` because the `get_comment()` function has a filtered return value, so a plugin could potentially substitute the comment with something completely different (although they probably wouldn't).

adamsilverstein replied:
Done in https://github.com/WordPress/wordpress-develop/pull/10489/commits/e74d3cfaed1130f401237c0a378a9d49ec1007a6

adamsilverstein committed:
Merge branch 'trunk' into fix/notification-filter-default

adamsilverstein committed:
Update src/wp-includes/comment.php

Co-authored-by: Weston Ruter <westonruter@gmail.com>

adamsilverstein committed:
Use comment object ID after get_comment() call

The get_comment() return value is filtered, so reassign
$comment_id from the comment object to ensure consistency
between the property checks and subsequent function calls.
Also removes a redundant null check on $comment.

adamsilverstein committed:
Merge branch 'trunk' into fix/notification-filter-default

adamsilverstein committed:
Merge branch 'trunk' into fix/notification-filter-default

adamsilverstein committed:
Merge branch 'trunk' into fix/notification-filter-default

adamsilverstein committed:
Use elseif and correct the @since version for the 7.1 milestone

adamsilverstein committed:
Add tests proving the notify_post_author filter has full control

The filter override and default-value tests fail against trunk, where the
approval check runs after the filter and silently vetoes a true result for
unapproved comments.

adamsilverstein committed:
Fix the invalid-comment test to avoid the global comment fallback

The test passed 0 as an invalid comment ID, but get_comment() treats any
empty ID as a request for the global $comment, which earlier tests in
the same file leave populated - so the instanceof guard never triggered
and the filter fired. Use a non-existent ID instead, which bypasses the
fallback and exercises the early return. Test-only change.

adamsilverstein committed:
Merge branch 'trunk' into fix/notification-filter-default

adamsilverstein committed:
Merge branch 'trunk' into fix/notification-filter-default

adamsilverstein committed:
Cast the comment ID to int before passing it to the filter

WP_Comment::$comment_ID is a numeric string, while both hooked call sites
previously delivered an int and the filter docblock documents the param as
int. Casting preserves strict comparisons in existing filter callbacks and
keeps the documented type accurate.

adamsilverstein committed:
Document the 7.1 behavior changes to wp_new_comment_notify_postauthor()

The approval check moving ahead of the notify_post_author filter and the
early return for invalid comment IDs are both observable to extenders, so
they need @since 7.1.0 entries on the function as well as the filter. Also
moves the function description above the @since tag per the inline
documentation standards.

adamsilverstein committed:
Add tests for the unapproved note and comments_notify defaults

Protects the two remaining default paths of the reordered logic: an
unapproved note must still default to notifying (the note carve-out), and
an approved comment's default must follow the comments_notify option.

adamsilverstein committed:
Merge branch 'trunk' into fix/notification-filter-default

westonruter committed:
Tidy up assertion logic by reusing MockAction and to pass PHPStan issues

westonruter commented on src/wp-includes/comment.php:
```diff
@@ -2462,25 +2462,42 @@ function wp_new_comment_notify_moderator( $comment_id ) {
+	$is_note    = ( 'note' === $comment->comment_type );
 
-	$maybe_notify = $is_note ? get_option( 'wp_notes_notify', 1 ) : get_option( 'comments_notify' );
+	// By default, only notify for approved comments and notes.
```
This comment is hard to connect with the conditions below. It says to only notify for approved comments and notes, but it doesn't seem to be doing that. It seems to not notify if a comment is not approved and it is not a note. But if it _is_ a note, then it seems to notify by default given `get_option( 'wp_notes_notify', 1 )`.

adamsilverstein replied:
Good catch, thanks - the branch order made the comment hard to follow. Reordered to handle notes first so each branch maps directly to the comment: notes are eligible regardless of approval status (per `wp_notes_notify`), other comments only once approved (per `comments_notify`). Fixed in 8bc0b7a9d4.

westonruter commented on src/wp-includes/comment.php:
```diff
@@ -2462,25 +2462,42 @@ function wp_new_comment_notify_moderator( $comment_id ) {
+	if ( '1' !== $comment->comment_approved && ! $is_note ) {
+		$maybe_notify = false;
+	} elseif ( $is_note ) {
+		$maybe_notify = get_option( 'wp_notes_notify', 1 );
```
```suggestion
		$maybe_notify = (bool) get_option( 'wp_notes_notify', 1 );
```

westonruter commented on src/wp-includes/comment.php:
```diff
@@ -2462,25 +2462,42 @@ function wp_new_comment_notify_moderator( $comment_id ) {
+	} elseif ( $is_note ) {
+		$maybe_notify = get_option( 'wp_notes_notify', 1 );
+	} else {
+		$maybe_notify = get_option( 'comments_notify' );
```
```suggestion
		$maybe_notify = (bool) get_option( 'comments_notify' );
```

adamsilverstein committed:
Apply suggestions from code review

Co-authored-by: Weston Ruter <westonruter@gmail.com>

adamsilverstein committed:
Clarify the default notification logic branches

Handle notes first so each branch maps directly to its documented
behavior: notes are eligible regardless of approval status, other
comments only once approved.
